### ISSUE TITLE
antsRegistrationSyNQuick.sh complains about bad arithmetic operator.

### ISSUE THREAD

**Issue by toomanycats** [ANTsX/ANTs#1809]
### Operating system and version

NAME="CentOS Linux" 
VERSION="7 (Core)" 
ID="centos" 
ID_LIKE="rhel fedora" 
VERSION_ID="7" 
PRETTY_NAME="CentOS Linux 7 (Core)" 

### CPU architecture

x86_64 (PC, Intel Mac, other Intel/AMD)

### ANTs code version

ants-2.5.0

### ANTs installation type

Other (please specify below)

### Summary of the problem

I cannot remember if I built ATNS or downloaded binaries, sorry.

I get the following warning:

```
WARNING /opt/ants/ants-2.5.0/bin/antsRegistrationSyNQuick.sh: line 447: ANTSVERSION=ants-2.5.0: syntax error: invalid arithmetic     operator (error token is ".5.0")
```

### Commands to reproduce the problem.

I create this script with Python and run it with `subprocess`:
```
export OMP_NUM_THREADS=1 && export ANTSVERSION=ants-2.5.0      
antsRegistrationSyNQuick.sh -d 3 \
                            -m /home/foo/pijp-pet/pet_templates/Template_FBP_all.nii.gz \
                            -f /m/Dev/ADNIDII-PET/pijp-pet/ADNIDII-0-3028-1b12/ADNIDII-0-3028-1b12.PET-Amy_moco_ave.nii.gz \
                            -t a \
                            -y 1 \
                            -o /m/Dev/ADNIDII-PET/pijp-pet/ADNIDII-0-3028-1b12/ADNIDII-0-3028-1b12.PET-Amy.template2Pet
```

### Output of the command with verbose output.

There is no verbose option. Further more, I suspect the help prints to `stderr` as I cannot pipe to `less` or `grep` the help.

### Data to reproduce the problem

Not sure data is required.

**Comment by gdevenyi** [comment]
> There is no verbose option. Further more, I suspect the help prints to stderr as I cannot pipe to less or grep the help.

You can do `|& less` to merge them into a single stdout stream to catch everything

**Comment by gdevenyi** [comment]
> I create this script with Python and run it with subprocess:

Please share exactly how you achieve the invocation in python

**Comment by toomanycats** [comment]
```
    def coreg_template_to_pet(self):
        # NOTE: `-m` --> "moving"
        # NOTE: `-f` --> "fixed"
        cmd = textwrap.dedent(f"""\
            export OMP_NUM_THREADS=1 && export ANTSVERSION=ants-2.5.0
            antsRegistrationSyNQuick.sh -d 3 \\
                                        -m {self.template} \\
                                        -f {self.pet_ave} \\
                                        -t a \\
                                        -y 1 \\
                                        -o {self.template_2_ave}""")

        self._run_cmd(cmd, 'reg_template_to_PET')

        # Find the xfm file and rename it.
        # ADNIDII-0-3028-1b12.PET-Amy.2Pet0GenericAffine.mat
        match = glob(self.template_2_ave + '0GenericAffine.mat')
        if len(match) != 1:
            raise ProcessingError("Failed to glob for affine matrix file.")
        os.rename(match[0], self.template_2_ave_xfm)

        # Find transformed template and rename.
        # ANTS will name it `self.template_2_ave + Warped.nii.gz`
        match = glob(self.template_2_ave + "Warped.nii.gz")
        if len(match) != 1:
            raise ProcessingError("Failed to glob for warped template file.")
        os.rename(match[0], self.template_2_ave + ".nii.gz")
```

**Comment by gdevenyi** [comment]
Thanks, have you isolated this outside of a python subprocess call?

**Comment by toomanycats** [comment]
> > There is no verbose option. Further more, I suspect the help prints to stderr as I cannot pipe to less or grep the help.
> 
> You can do `|& less` to merge them into a single stdout stream to catch everything

Did you mean `&>` ? That didn't seem to work though. I thought I understood redirects but seems I like never do.

**Comment by cookpa** [comment]
I think you should use the `env` argument for subprocess, rather than trying to put them into `cmd`.

I also think the args to `cmd` should be a list, not a string. 

For example,

```
subprocess.run(['antsRegistrationSyNQuick.sh', '-d', '3', '-f', fixed_image, '-m', moving_image, -o my_output_prefix])
```

**Comment by gdevenyi** [comment]
Also, please be aware that `exports` with subprocess are not best practice for setting environment settings, see https://docs.python.org/3/library/subprocess.html for the details on the `env` option.

**Comment by cookpa** [comment]
You can set threading through the `-n` option, or use the environment variable `ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS`. ANTs doesn't use OMP threads

**Comment by gdevenyi** [comment]
> Did you mean &> ? That didn't seem to work though. I thought I understood redirects but seems I like never do.

`|&` merges `stderr` and `stdout` into a single pipe to the next process in line such as `less`

**Comment by toomanycats** [comment]
Found the issue...it's not your fault.

It's our stupid version wrapper:
```
$ PrintHeader ADNIDII-0-3028-1b12.PET-Amy_moco_ave.nii.gz 2
ANTSVERSION=ants-2.5.0                                                                                                                                      192x192x89                   
```

**Comment by cookpa** [comment]
Ah that'd do it. If you need the local variable for the call, I think you can do it without the export, eg

```
ANTSVERSION=ants-2.5.0  antsRegistrationSyNQuick.sh -d 3 \
...
```

**Comment by toomanycats** [comment]
Our system to allowing multiple versions of software uses a wrapper that tends to `echo` at the worst times.
This threw off the scripts parsing of the image size.

**Comment by toomanycats** [comment]
> Ah that'd do it. If you need the local variable for the call, I think you can do it without the export, eg
> 
> ```
> ANTSVERSION=ants-2.5.0  antsRegistrationSyNQuick.sh -d 3 \
> ...
> ```

Long story...the way our lab handles multiple versions of the same software. The fix is the use `Modules`.

**Comment by gdevenyi** [comment]
Should you have a need to setup modules for your lab, https://github.com/gdevenyi/NeuroAnsible/

**Comment by toomanycats** [comment]
@gdevenyi Your ansible playbook is substantial. The only thing it doesn't do, is handle multiple version simultaneously.
I usually build whatever, ANTS, AFNI, FS on a commonly shared storage then mount it everywhere.

There's a lot of nice stuff in there though. Now we're moving to Alma 9 things should get easier. 
Building AFNI's R library for CentOS 7 was total PITA.

**Comment by gdevenyi** [comment]
> The only thing it doesn't do, is handle multiple version simultaneously.

It does, I just haven't done one yet, and I wish I had a nicer templating system for that

> I usually build whatever, ANTS, AFNI, FS on  shared storage then mount it everywhere.

Indeed that's what I do for the system I manage  at work, but I wanted to provide something for people to use that could work easily on a single system.

I have deployed this module system across a shared deployment over NFS as well. I should make adjust the tag system so that it's easier to make that happen without hackery.


> There's a lot of nice stuff in there though. Now we're moving to Alma 9 things should get easier. 
> Building AFNI's R library for CentOS 7 was total PITA.

Thanks.

With the availability of binaries from upstream in many cases now, I've decided it's no longer worth the effort to build all the software from scratch unless I absolutely have to.

If I did want to do a module system building from source, I would use https://spack.readthedocs.io/en/latest/ and contribute missing recipes there for more neuroimaging software.
